File: longestIncreasingPath.py
class Solution:
    def longestIncreasingPath(self, matrix: List[List[int]]) -> int:
        
        if not matrix:
            return 0

        j = len(matrix[0])
        n = len(matrix)
        max_path = 0

        dp = [[0] * j for _ in range(n)]
    

        def recurs(i, k):
            if dp[i][k] : 
                return dp[i][k]

            best = 1
            for ji, jk in ((1,0), (0,1), (-1, 0), (0, -1)):
                ni, nj = i+ji, k+jk
                if 0 <= ni < n and 0 <= nj < j and matrix[ni][nj] > matrix[i][k]:
                    best = max(best, 1 + recurs(i+ji, k+jk))
            dp[i][k] = best
            return dp[i][k]
        
        for i in range(n):
            for k in range(j):
                max_path = max(max_path, recurs(i, k))

        return max_path

File: longestIncreasingPath.rs
impl Solution {
    pub fn longest_increasing_path(matrix: Vec<Vec<i32>>) -> i32 {
        let rows = matrix.len();
        if rows == 0 || matrix[0].is_empty() {
            return 0;
        }
        let cols = matrix[0].len();
        let dirs = [(1,0),(-1,0),(0,1),(0,-1)];
        let mut dp = vec![vec![0; cols]; rows];

        fn dfs(
            i: usize,
            j: usize,
            matrix: &Vec<Vec<i32>>,
            dp: &mut Vec<Vec<i32>>,
            dirs: &[(i32, i32)],
        ) -> i32 {
            if dp[i][j] != 0 {
                return dp[i][j];
            }
            let mut best = 1;
            let rows_i32 = matrix.len() as i32;
            let cols_i32 = matrix[0].len() as i32;

            for &(di, dj) in dirs {
                let ni = i as i32 + di;
                let nj = j as i32 + dj;
                if (0..rows_i32).contains(&ni)
                    && (0..cols_i32).contains(&nj)
                    && matrix[ni as usize][nj as usize] > matrix[i][j]
                {
                    best = std::cmp::max(
                        best,
                        1 + dfs(ni as usize, nj as usize, matrix, dp, dirs),
                    );
                }
            }

            dp[i][j] = best;
            best
        }

        let mut ans = 0;
        for i in 0..rows {
            for j in 0..cols {
                ans = std::cmp::max(ans, dfs(i, j, &matrix, &mut dp, &dirs));
            }
        }
        ans
    }
}

